js/src/components/adaptive-form/adaptive-form.js

```diff
@@ -63,10 +67,13 @@ function AdaptiveForm( { onSubmit, children, ...props }, ref ) {
 		...formRef.current,
 	} ) );
 
-	// Enhancements related.
-	// Additional enhancements to make AdaptiveForm have more useful or reusable features.
-	// It could also be the playground of the practical instances before pushing them upstream.
+	/* === Start of enhancement-related codes === */
+
 	const isMounted = useIsMounted();
+
+	// Add `isSubmitting` and `isSubmitted` states for facilitating across multiple layers of
+	// component controlling, such as disabling inputs or buttons. They are composed in a tuple
+	// to reduce the number of triggered re-rendering.
```

Comment:
Ok, reading it now again I understand now this code.

The first element in the tuple is the `isSubmitting` state and the second is the `isSubmitted ` state. The first time I didn't read it properly. I assumed somehow the second line was another different state (maybe because I read it fast). 

I think it will be clearer naming the submission statuses instead of using tuples. 

```jsx
const [ submission, setSubmission ] = useState( null );
const isSubmitting = submission === SUBMITTING ;
const isSubmitted = submission === SUBMITTED ;

if ( onSubmit ) {
		props.onSubmit = async function ( values ) {
			setSubmission( SUBMITTING );
			await onSubmit.call( this, values, adapterRef.current.submitter );
			if ( isMounted() ) {
				setSubmission( SUBMITTED );
				adapterRef.current.submitter = null;
			}
		};
	}
```

Reply:
I still don't get why `isMounted` is related to the submission of the form. Maybe I'm missing some of the internals tho...

Reply:
My bet is that if user clicks submit, then the form is on isSubmitting, If the user leves this page between that moment and the completion of the onSubmit call , the component is unMounted. So `setSubmission( SUBMITTED );`should not be called. This isMounted() prevents that case. Is that right?

Reply:
> I think it will be clearer naming the submission statuses instead of using tuples.

Thanks for the nice suggestion. 👍  I changed to use it in https://github.com/woocommerce/google-listings-and-ads/pull/1794/commits/1c80e7d8761c7138c66bd1664abc0b9c29be574a.

> My bet is that if user clicks submit, then the form is on isSubmitting, If the user leves this page between that moment and the completion of the onSubmit call , the component is unMounted. So `setSubmission( SUBMITTED );`should not be called. This isMounted() prevents that case. Is that right?

Yes, It's indeed to deal with this situation. Otherwise, the following warning would appear even though it won't cause an actual memory problem here.

![image](https://user-images.githubusercontent.com/17420811/204789018-3c9a49df-744d-40b6-8e89-0f5ad2ec479a.png)

Reply:
A related discussion about `useIsMounted`: https://github.com/woocommerce/google-listings-and-ads/pull/1007#discussion_r708838891